src/trace/trace.c

```diff
@@ -96,44 +101,145 @@ static inline void mtrace_event(const char *data, uint32_t length)
 	platform_shared_commit(trace, sizeof(*trace));
 }
 
+#if CONFIG_TRACE_FILTERING_VERBOSITY
 /**
- * \brief Runtime trace filtering
+ * \brief Runtime trace filtering based on verbosity level
  * \param lvl log level (LOG_LEVEL_ ERROR, INFO, DEBUG ...)
  * \param uuid uuid address
  * \return false when trace is filtered out, otherwise true
  */
-static inline bool trace_filter_pass(uint32_t lvl,
-				     const struct tr_ctx *ctx)
+static inline bool trace_filter_verbosity(uint32_t lvl, const struct tr_ctx *ctx)
 {
 	/* LOG_LEVEL_CRITICAL has low value, LOG_LEVEL_VERBOSE high */
 	return lvl <= ctx->level;
 }
+#endif /* CONFIG_TRACE_FILTERING_VERBOSITY */
+
+#if CONFIG_TRACE_FILTERING_ADAPTIVE
+struct recent_log_entry {
+	uint32_t entry_id;
+	uint64_t message_ts;
+#if CONFIG_TRACE_RECENT_PROLONG
+	uint64_t first_suppression_ts;
+#endif /* CONFIG_TRACE_RECENT_PROLONG */
+	uint32_t suppress_count;
+};
+
+static struct recent_log_entry recent_entries[CONFIG_TRACE_RECENT_ENTRIES_COUNT];
+static struct recent_log_entry entry_to_emit;
+
+static inline void prepare_to_emit(struct recent_log_entry *entry)
+{
+	if (entry->suppress_count)
+		memcpy_s(&entry_to_emit, sizeof(entry_to_emit), entry, sizeof(*entry));
+	memset(entry, 0, sizeof(*entry));
+}
+
+static inline void emit_recent_entry(void)
+{
+	if (!entry_to_emit.suppress_count)
+		return;
+
+	_log_message(trace_log_unfiltered, false, LOG_LEVEL_INFO, _TRACE_INV_CLASS, &dt_tr,
+		     _TRACE_INV_ID, _TRACE_INV_ID, "Suppressed %u similar messages: %s",
+		     entry_to_emit.suppress_count, entry_to_emit.entry_id);
```

Comment:
%s formatting was recently turned off in logger: https://github.com/thesofproject/sof/pull/3472
What formatting specifier should I use, to handle the special case of printing *contents* of the log entry?
Should I check for %s, and if argument 'might' be log_entry, then process it that way, and return error otherwise (as is done now)? Or use a completely new specifier and create a special handler in logger?

Reply:
We already have dedicated `%pU` specifier for UUID, `%ps` or `%pT` could constructed using the same schema as for UUID. `%s` is reserved for non-supprted feature - runtime string printing, using this common known specifier for other purposes is misleading.

Reply:
This is going to report to the user which log was suppressed and now released. Yes, this requires patch to the logger, which I will send as separate PR, once a suitable format specifier is decided on.

Reply:
I've added the logger part to this PR, to hopefully close this feature up easier.